### PR TITLE
Issue 3433 - additional utility scripts and an introduction to a scheduler

### LINKED ISSUE #3433: Additional utility/cron scripts
### Description

Here is a list of some useful utility scripts that form part of a well run installation of io

#### Periodic share directory cleaner
_Scheduled?: daily_
To allow for time to investigate failed model processing, the current bouncer process leaves the uploaded models on the folder that is shared with bouncer. However for data retention and storage reasons, it's sensible to clear those down. We currently run a script externally to clear down the $SHARED_DIR/queue to maintain it at 14 days of age of file.

#### Remove Unverified Users
_Scheduled?: daily_
Exists currently as a [separate mongo js](https://github.com/3drepo/3drepo.io/blob/master/scripts/dbMaintenance/removeUnverifiedExpiredUsers.js) but could be rolled into a utility script for ease of management and implementation. It has also been copied into the helm chart for installation currently by using the contents of the script.

#### Reset processing flag
_Scheduled?: no_
Occasionally there may be an error with model processing that requires the flag to be reset on a particular model or teamspace, or the entire system. A script exists currently but it isn't possible to target a teamspace or particular model

#### license verification and information check
_Scheduled?: no_
Whilst io is open source and does not require a license to operate, 3D Repo does offer a licensed system for the 3drepobouncer/3drepounity system. It is possible to [add your license key to the config ](https://github.com/3drepo/3drepo.io/pull/3042) for better reporting, so if present it could be useful to check the expiry details and permissions set.

<details><summary>Belongs in an endpoint thus out of scope</summary>
-  Queue Monitor

We are exporting the queue stats to cloudwatch using an [external system](https://github.com/3drepo/DevOps/tree/master/docker/3drepoqueuemonitor) currently, and this could be rolled into IO.

</details>

<details><summary>Belongs to a devops script suite thus won't implement</summary>
- General Utility option to loop with a delay on long running tasks

Occasionally we wish to run batches of commands slowly over time as to not overload the system, so a configuration option to have loops operate at an interval could be useful.

- rabbitMQ initialisation

Currently we init the rabbitMQ queues externally to an IO installation but in the case of a local installation for development or small installations it could be useful to have the commands required to bootstrap a persistent rabbitmq installation.

- MongoDB initialisation

In the past there were mongo scripts created to create custom roles that give tight security for the minimal amount of operations required for IO to operate in a secure manner. In local deployments these steps are missed, so developed under accounts with higher permissions than required. We could revisit this to make it easier for new installations to be configured securely.
</details>

### Goals
- [x] Review suggestions and decide what to implement
- [x] Implement additional scripts
- [ ] Implement a way to run a scheduled script

### Tasks
- [x] Write new utility scripts
  - [x] Periodic share directory cleaner
  - [x] Remove Unverified Users
  - [x] Reset processing flag
  - [x] License verification and information check
- [x] Implement a way to run scheduled scripts
  - [x] Add email notification if scheduled scripts failed

### PR DESCRIPTION
**Pull request by carmenfan**
This fixes #3433 
#### Description
- Adding the following scripts
  - `checkLicense` - Attempts to activate the configured 3drepo license and dumps out the license information
  - `cleanUpSharedDir` - Remove files older than 14 days(default) in the shared directory
  - `resetProcessingFlags` - Resets the processing flag on all/certain teamspace/models
  - `removeUnverifiedUsers` - Remove all unverified users with an expired verification token

- Added a new action on yarn/npm to run scheduled tasks (read [README.md](https://github.com/3drepo/3drepo.io/blob/ISSUE_3433/backend/src/scripts/utility/README.md) for details)

#### Test cases
- additional scripts should run as expected
- scheduler should run as expected

### CODE REVIEW

**Comment by carmenfan** [comment]
@keifgwinn  tagging you to review this one first to make sure it fits your workflow. 2 things to check here:

1. whether the additional scripts work and whether you need more configurable values for any of them.
2. whether the scheduler would fit the way we run deployments, and whether anything else needs to be more configurable.

Also an additional question: if the scheduler errored out, do we want to send an email to devops (like failed model processing) or whether you'd just hook up an external alarm system, or it should only send an email if we configured it to do so. Either way I think we want to know if it failed so it's just how we want to alert ourselves.

File: backend/src/scripts/utility/modelProcessing/cleanUpSharedDir.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		}
+	}))).flat();
+
+	logger.logInfo(`${filesToRemove.length} files/directories deleted.`);
+};
+
+const genYargs = /* istanbul ignore next */(yargs) => {
+	const commandName = Path.basename(__filename, Path.extname(__filename));
+	const argsSpec = (subYargs) => subYargs.option('threshold', {
+		describe: 'Files existed before the threshold(in days) will be removed',
```
Comment (keifgwinn) [inline backend/src/scripts/utility/modelProcessing/cleanUpSharedDir.js:70]: 'Files that were created before the threshold (in days) will be removed'

is my suggestion for the description

**Comment by keifgwinn** [comment]
Error handling on the params being out of range has the stack trace

`yarn run-script cleanUpSharedDir --threshold 0
````
yarn run-script cleanUpSharedDir

Remove any file processing data from file share that are x days old

Options:
  --help       Show help                                                                                                                                                                                                        [boolean]
  --version    Show version number                                                                                                                                                                                              [boolean]
  --threshold  Files existed before the threshold(in days) will be removed                                                                                                                                         [number] [default: 14]

Error: Threshold must be at least 1
    at run (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\modelProcessing\cleanUpSharedDir.js:35:9)
    at Object.handler (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\modelProcessing\cleanUpSharedDir.js:77:13)
    at C:\Dev\3drepo\3drepo.io\backend\node_modules\yargs\build\index.cjs:1:9054
    at j (C:\Dev\3drepo\3drepo.io\backend\node_modules\yargs\build\index.cjs:1:4931)
    at M.applyMiddlewareAndGetResult (C:\Dev\3drepo\3drepo.io\backend\node_modules\yargs\build\index.cjs:1:9023)
    at M.runCommand (C:\Dev\3drepo\3drepo.io\backend\node_modules\yargs\build\index.cjs:1:7206)
    at [runYargsParserAndExecuteCommands] (C:\Dev\3drepo\3drepo.io\backend\node_modules\yargs\build\index.cjs:1:56715)
    at Jt.parse (C:\Dev\3drepo\3drepo.io\backend\node_modules\yargs\build\index.cjs:1:38938)
    at Object.<anonymous> (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\index.js:62:3)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
```

**Comment by carmenfan** [comment]
> 

@keifgwinn is that a problem? It's always been like that on errors, for e.g. zombie removal script (which was done a few back)
image


This looks like it's handled by Yargs actually. It's not falling through to any of my catches

**Comment by keifgwinn** [comment]
For unknown errors sure but this is expected bad value so seeing a stack trace implies we have a bigger problem that the code doesn't understand what I've told it,

**Comment by carmenfan** [comment]
> For unknown errors sure but this is expected bad value so seeing a stack trace implies we have a bigger problem that the code doesn't understand what I've told it.
> 
> I didn't like the behaviour before either ;)

How badly do you want this?

Basically when we run into an error, we throw an exception (pretty standard practice). Yargs catches the exception and handle it this way.

In order to do what you want, we have to either:

1. create our own Error class that inherits the standard error class with an additional flag to denote whether we want to show a stack trace or not and handle that through the bespoke error handler
2. create different error classes and check the instance of objects to determine whether we show a stack trace or not

TLDR;. a substantial amount of effort on something you can probably just turn a blind eye on, also relies on future script writes to remember the differences and do the right thing 😆 

or... we add a verbosity flag so the trace only shows if you set that to true - which will be less effort, but that's still subject to overriding the error handler

**Comment by keifgwinn** [comment]
adding in a valid choices option however means out of bounds gets handled by yargs. 

`		choices: Array.from({ length: DEFAULT_THRESHOLD }, (v, k) => k + 1),` and testing gives

image

the downside of this approach is it limits the input and also writes it out if we use a large one like 90 or 999, but allows people to select sooner than 14 days if they really want to get up to 1 day old but not break things at 0.

**Comment by carmenfan** [comment]
> it's not about the max, I was just poking through the yargs docs trying to find a way to make it check acceptable ranges on inputs. My point was it didn't trigger a stack trace as yargs handled it.

We have predictable errors within run that may not require a stack trace either (e.g. teamspace not found)

if you care about stack trace then I would still think the best options are as I suggested https://github.com/3drepo/3drepo.io/pull/3966#issuecomment-1430092667

**Comment by carmenfan** [comment]
@keifgwinn if you're bothered about the stack trace i'll just code in the verbosity flag solution, I don't think it's worth delving into and spend more time on it

File: backend/src/scripts/utility/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    // This will clean up any files within the shared directory older than 14 days (using default params)
+    { "name": "cleanUpSharedDir" } 
+  ],
+  "weekly": [
+    // This will remove incomplete revisions older than 5 days (instead of the 14 days default)
+    {"name": "removeIncompleteRevisions", params: [5]} 
+  ],
+  "monthly": [
+    // This will print out a report of all active licenses to C://dataOutput
+    {"name": "allActiveLicenses", params: ["C://dataOutput"]}
```
[the remark this answers is not in the document]
Comment (keifgwinn) [reply backend/src/scripts/utility/README.md:45]: or in fact all environment expansion would be good. For simplicity we could standardise on %% windows style env so it's got a clear delimiter. I wanted to use something like %TEMP% for a folder path and %CURRENTTIME% for a logfile name.

File: backend/src/scripts/utility/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+An example:
+````js
+{
+  "daily": [
+    // This will clean up any files within the shared directory older than 14 days (using default params)
+    { "name": "cleanUpSharedDir" } 
+  ],
+  "weekly": [
+    // This will remove incomplete revisions older than 5 days (instead of the 14 days default)
+    {"name": "removeIncompleteRevisions", params: [5]}
```
[the remark this answers is not in the document]
Comment (keifgwinn) [reply backend/src/scripts/utility/README.md:41]: or that ./backend/src/scripts/utility/scheduler.config.json exists as a default template

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/scripts/utility/README.md:41]: > also a comment above that comments should be removed from the file

Comments are just there to explain, I don't intend for this to be a valid json.

And yes ` ./backend/src/scripts/utility/scheduler.config.json` is the default template if you pass in a config path, also serving as an example

**Comment by carmenfan** [comment]
> @keifgwinn if you're bothered about the stack trace i'll just code in the verbosity flag solution, I don't think it's worth delving into and spend more time on it

@keifgwinn 

image

**Comment by keifgwinn** [comment]
```
❯ .\yarn.ps1 run-scheduled-task --config .\3drepo.io\config\development\schedule.json
2023-02-16T10:27:43.125Z [info] [APP]   ======================== Daily tasks ======================== 
2023-02-16T10:27:43.127Z [info] [APP]   [1/3] allActiveLicenses with ["%TEMP%//report.csv"]...
2023-02-16T10:27:43.197Z [info] [APP]           -localsignup01 
2023-02-16T10:27:43.210Z [info] [APP]           -ssotest1
2023-02-16T10:27:43.214Z [info] [APP]           -ssotest2 
2023-02-16T10:27:43.217Z [error] [APP]  string.slice is not a function
keifg@KEIF  C:\Dev\3drepo\local   ISSUE_92 ≢                                                                                                                                          [10:27]
❯ .\yarn.ps1 run-scheduled-task --config .\3drepo.io\config\development\schedule.json -v
2023-02-16T10:28:03.080Z [info] [APP]   ======================== Daily tasks ======================== 
2023-02-16T10:28:03.082Z [info] [APP]   [1/3] allActiveLicenses with ["%TEMP%//report.csv"]...
2023-02-16T10:28:03.142Z [info] [APP]           -localsignup01 
2023-02-16T10:28:03.150Z [info] [APP]           -ssotest1
2023-02-16T10:28:03.154Z [info] [APP]           -ssotest2
2023-02-16T10:28:03.158Z [error] [APP]  string.slice is not a function 
TypeError: string.slice is not a function
    at stringSplice (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utils.js:66:24)
    at Utils.parsePath (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utils.js:83:10)
    at Object.run [as allActiveLicenses] (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\teamspaces\allActiveLicenses.js:55:32)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async runScripts (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\scheduler.js:77:3)
    at async run (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\scheduler.js:112:2)
```

-v works nicely =)

File: backend/src/scripts/utils.js [hunk 4]
```diff
@@ -55,4 +62,39 @@ Utils.getCollectionsEndsWith = async (teamspace, str) => {
 	return collections.filter(({ name }) => name.endsWith(str));
 };
 
+const stringSplice = (string, index, length, replacement) => {
+	const before = string.slice(0, index);
```
Comment (keifgwinn) [inline backend/src/scripts/utils.js:66]: using this config I got 
```

TypeError: string.slice is not a function
    at stringSplice (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utils.js:66:24)
    at Utils.parsePath (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utils.js:83:10)
    at Object.run [as allActiveLicenses] (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\teamspaces\allActiveLicenses.js:55:32)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async runScripts (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\scheduler.js:77:3)
    at async run (C:\Dev\3drepo\3drepo.io\backend\src\scripts\utility\scheduler.js:112:2)
```

```
{
  "daily": [
    {"name": "allActiveLicenses", "params": ["%TEMP%//report.csv"]}, 
    { "name": "cleanUpSharedDir", "params": [5] } ,
    { "name": "viewLicense","params": ["localsignup01"] }
  ],
  "weekly": [
    {"name": "removeIncompleteRevisions", "params": [5]} 
  ],
  "monthly": [
  ]
}

```

Comment (carmenfan) [reply backend/src/scripts/utils.js:66]: image

**Comment by carmenfan** [comment]
> Also an additional question: if the scheduler errored out, do we want to send an email to devops (like failed model processing) or whether you'd just hook up an external alarm system, or it should only send an email if we configured it to do so. Either way I think we want to know if it failed so it's just how we want to alert ourselves.


@keifgwinn  any comments regarding this?

File: backend/src/scripts/utility/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+An example:
+````js
+{
+  "daily": [
+    // This will clean up any files within the shared directory older than 14 days (using default params)
+    { "name": "cleanUpSharedDir" }
+  ],
+  "weekly": [
+    // This will remove incomplete revisions older than 5 days (instead of the 14 days default)
+    {"name": "removeIncompleteRevisions", params: [5]}
```
Comment (keifgwinn) [inline backend/src/scripts/utility/README.md:41]: another quote on params

**Comment by keifgwinn** [comment]
> Also an additional question: if the scheduler errored out, do we want to send an email to devops (like failed model processing) or whether you'd just hook up an external alarm system, or it should only send an email if we configured it to do so. Either way I think we want to know if it failed so it's just how we want to alert ourselves.

I think send an email if configured to do so is the most user friendly way. Ideally I would probably monitor it externally but not everyone will have access to that, and it's also easier for me to set up right away now.

**Comment by carmenfan** [comment]
@keifgwinn  all fixed (🤞 )

there is now a emailOnFailure flag that can be set/unset on the config https://github.com/3drepo/3drepo.io/blob/ISSUE_3433/backend/src/scripts/utility/README.md#scheduler-configurations

File: backend/src/scripts/utility/README.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
+# Utility Suite
+The utility script consist of a library of helpful commands to extract or update the data within the application.
+
+To see the list of commands and how to run them, run `yarn run-script --help` on the [backend root folder](../../../)
+
+## Scheduler
+[The scheduler](./scheduler.js) is an executable script. Depending on the [configuration](scheduler.config.js), it will execute certain script(s) within the utility suite. It is designed to be hooked onto a job scheduler (e.g. `cron`) to be executed on a daily basis.
```
Comment (Charence) [inline backend/src/scripts/utility/README.md:7]: The configuration link doesn't seem to work

Comment (carmenfan) [reply backend/src/scripts/utility/README.md]: Fixed

File: backend/src/scripts/utility/index.js [hunk 7]
```diff
@@ -59,7 +61,15 @@ const parser = populateCommands(Yargs(hideBin(process.argv)))
...
+		handleErrorBeforeExit(msg ?? err);
+	})
 	.parse();
 
-// eslint-disable-next-line no-console
-Promise.resolve(parser).catch(console.error).finally(process.exit);
+Promise.resolve(parser).catch(handleErrorBeforeExit).finally(() => {
+	// eslint-disable-next-line no-process-exit
+	process.exit();
+});
```
Comment (Charence) [inline backend/src/scripts/utility/index.js:75]: @carmenfan Running `yarn run-script` without specifying a script now gives an error:
```
yarn run-script
yarn run v1.22.5
$ cross-env NODE_CONFIG_DIR=../config/remote node ./src/scripts/utility/index
Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
2023-03-06T12:42:50.405Z [error] [APP] 	Not enough non-option arguments: got 0, need at least 1 
error Command failed with exit code 1.
```

Comment (carmenfan) [reply backend/src/scripts/utility/index.js:75]: I don't see a problem with this. You didn't give it any command so it didn't know what to do. It shoulds the options so the user knows to `--help` to proceed

@keifgwinn  do you care?

Comment (keifgwinn) [reply backend/src/scripts/utility/index.js:75]: yes but not enough to stop this pull request from going through as the other functions are more useful

**Review by keifgwinn — APPROVED** [review]
